From c589d3f665e576cda2d0c28209be00579d9ebca5 Mon Sep 17 00:00:00 2001
From: Iain Lane <iain.lane@canonical.com>
Date: Thu, 28 Jan 2016 13:35:32 +0000
Subject: [PATCH 2/2] Spawn Ubuntu's credentials panel instead of the GOA
 online-accounts one on Unity

Show the cog button to launch the control-center in the remaining
GtkGrid in this case.

(Backported to gnome-3-18, use the patches on the bug for 3.20)

https://bugzilla.gnome.org/show_bug.cgi?id=761259

---
 src/gcal-source-dialog.c | 51 ++++++++++++++++++++++++++++++++++++------------
 1 file changed, 39 insertions(+), 12 deletions(-)

Index: b/src/gcal-source-dialog.c
===================================================================
--- a/src/gcal-source-dialog.c
+++ b/src/gcal-source-dialog.c
@@ -39,6 +39,9 @@
   GtkWidget          *name_entry;
   GtkWidget          *notebook;
   GtkWidget          *online_account_main_grid;
+  GtkWidget          *online_accounts_settings_button;
+  GtkWidget          *other_calendars_main_grid;
+  GtkWidget          *other_calendars_main_frame;
   GtkWidget          *remove_button;
   GtkWidget          *stack;
   GtkWidget          *web_source_grid;
@@ -498,6 +501,29 @@
     gcal_manager_set_default_source (self->manager, self->old_default_source);
 }
 
+static void
+spawn_goa_with_args (gchar *action,
+                     gchar *arg)
+{
+  const gchar *desktop;
+
+  gchar *command[] = {"gnome-control-center", "online-accounts", action, arg, NULL};
+
+  /* On Ubuntu (Ubuntu Online Accounts), the panel is called 'credentials' */
+  desktop = g_getenv ("XDG_CURRENT_DESKTOP");
+  if (desktop && strstr (desktop, "Unity"))
+  {
+    command[1] = "credentials";
+  }
+
+  g_spawn_async (NULL, command, NULL, G_SPAWN_SEARCH_PATH | G_SPAWN_STDOUT_TO_DEV_NULL, NULL, NULL, NULL, NULL);
+}
+
+static void
+spawn_goa (void)
+{
+  spawn_goa_with_args (NULL, NULL);
+}
 
 /**
  * description_label_link_activated:
@@ -512,8 +538,7 @@
                                   gchar     *uri,
                                   gpointer   user_data)
 {
-  gchar *command[] = {"gnome-control-center", "online-accounts", NULL};
-  g_spawn_async (NULL, command, NULL, G_SPAWN_SEARCH_PATH, NULL, NULL, NULL, NULL);
+  spawn_goa ();
 
   return TRUE;
 }
@@ -653,14 +678,6 @@
     e_source_set_display_name (self->source, gtk_entry_get_text (GTK_ENTRY (self->name_entry)));
 }
 
-static void
-spawn_goa_with_args (gchar *action,
-                     gchar *arg)
-{
-  gchar *command[] = {"gnome-control-center", "online-accounts", action, arg, NULL};
-  g_spawn_async (NULL, command, NULL, G_SPAWN_SEARCH_PATH | G_SPAWN_STDOUT_TO_DEV_NULL, NULL, NULL, NULL, NULL);
-}
-
 /**
  * online_accounts_listbox_row_activated:
  *
@@ -1095,8 +1112,7 @@
 online_accounts_settings_button_clicked (GtkWidget *button,
                                          gpointer   user_data)
 {
-  gchar *command[] = {"gnome-control-center", "online-accounts", NULL};
-  g_spawn_async (NULL, command, NULL, G_SPAWN_SEARCH_PATH, NULL, NULL, NULL, NULL);
+  spawn_goa ();
 }
 
 static void
@@ -1798,6 +1814,14 @@
   desktop = g_getenv ("XDG_CURRENT_DESKTOP");
   if (desktop && strstr (desktop, "Unity"))
     {
+      gtk_container_remove (GTK_CONTAINER (self->online_account_main_grid),
+                            self->online_accounts_settings_button);
+      gtk_container_add (GTK_CONTAINER (self->other_calendars_main_grid),
+                         self->online_accounts_settings_button);
+      gtk_container_child_set (GTK_CONTAINER (self->other_calendars_main_grid),
+                               self->other_calendars_main_frame,
+                               "width", 3,
+                               NULL);
       gtk_widget_hide (self->online_account_main_grid);
     }
 
@@ -1882,6 +1906,9 @@
   gtk_widget_class_bind_template_child (widget_class, GcalSourceDialog, notification_label);
   gtk_widget_class_bind_template_child (widget_class, GcalSourceDialog, online_accounts_listbox);
   gtk_widget_class_bind_template_child (widget_class, GcalSourceDialog, online_account_main_grid);
+  gtk_widget_class_bind_template_child (widget_class, GcalSourceDialog, online_accounts_settings_button);
+  gtk_widget_class_bind_template_child (widget_class, GcalSourceDialog, other_calendars_main_grid);
+  gtk_widget_class_bind_template_child (widget_class, GcalSourceDialog, other_calendars_main_frame);
   gtk_widget_class_bind_template_child (widget_class, GcalSourceDialog, owncloud_stub_row);
   gtk_widget_class_bind_template_child (widget_class, GcalSourceDialog, remove_button);
   gtk_widget_class_bind_template_child (widget_class, GcalSourceDialog, settings_button);
